### PR TITLE
Add org.kde.ghostwriter

### PR DESCRIPTION
**Pull request by Aman9das**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [ ] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. [I was a maintainer of the previous flatpak, ownership could be handed to kde developers.
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission

### CODE REVIEW

**Comment by Aman9das** [comment]
bot, build org.kde.ghostwriter

**Comment by hfiguiere** [comment]
cc @flathub/kde

File: org.kde.ghostwriter.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  sources:
+  - type: git
+    url: https://github.com/commonmark/cmark.git
+    tag: 0.30.3
+- name: cmark-gfm
+  buildsystem: cmake-ninja
+  builddir: true
+  config-opts:
+  - "-DCMAKE_BUILD_TYPE=Release"
+  - "-DCMAKE_INSTALL_PREFIX=/app"
```
Comment (hfiguiere) [inline org.kde.ghostwriter.yml:34]: this is not necessary
```suggestion
```

File: org.kde.ghostwriter.yml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  sources:
+  - type: git
+    url: https://github.com/github/cmark.git
+    tag: 0.29.0.gfm.13
+- name: MultiMarkdown
+  buildsystem: cmake-ninja
+  builddir: true
+  config-opts:
+  - "-DCMAKE_BUILD_TYPE=Release"
+  - "-DCMAKE_INSTALL_PREFIX=/app"
```
Comment (hfiguiere) [inline org.kde.ghostwriter.yml:44]:
```suggestion
```
this is not necessary

File: org.kde.ghostwriter.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  sources:
+  - type: git
+    url: https://github.com/fletcher/MultiMarkdown-6.git
+    tag: 6.7.0
+- name: Pandoc
+  only-arches:
+  - x86_64
```
Comment (hfiguiere) [inline org.kde.ghostwriter.yml:51]:
```suggestion
```

File: org.kde.ghostwriter.yml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  - type: archive
+    only-arches: ["x86_64"]
+    url: https://github.com/jgm/pandoc/releases/download/3.1.6/pandoc-3.1.6-linux-amd64.tar.gz
+    sha256: 6eec0e0c85e97c90a02b9184c993d4bd848426fb3d7452ddc0d6014368c5e3fd
+  - type: archive
+    only-arches: ["aarch64"]
+    url: https://github.com/jgm/pandoc/releases/download/3.1.6/pandoc-3.1.6-linux-arm64.tar.gz
+    sha256: 6ca797f358b0cc04a3b9208ff1975a861bfa7110eafa881aa8c023ac4040e208
+- name: ghostwriter
+  buildsystem: cmake
```
Comment (hfiguiere) [inline org.kde.ghostwriter.yml:66]: if it's cmake, why the QMAKE stuff below?

File: org.kde.ghostwriter.yml [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  build-options:
+    env:
+    - QMAKEPATH=/app/lib
+  config-opts:
+  - QMAKE_INCDIR+=/app/include/QtWebEngine
+  - QMAKE_INCDIR+=/app/include/QtWebEngineCore
+  - QMAKE_INCDIR+=/app/include/QtWebEngineWidgets
+  sources:
+  - type: git
+    url: https://github.com/KDE/ghostwriter.git
```
Comment (hfiguiere) [inline org.kde.ghostwriter.yml:77]: aren't official KDE project pulled from tarballs from the KDE infra?

File: org.kde.ghostwriter.yml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+- "--filesystem=home"
+- "--device=dri"
+- "--env=QTWEBENGINE_DISABLE_SANDBOX=1"
+modules:
+- name: cmark
+  buildsystem: cmake-ninja
+  builddir: true
+  config-opts:
+  - "-DCMAKE_BUILD_TYPE=Release"
+  - "-DCMAKE_INSTALL_PREFIX=/app"
```
Comment (hfiguiere) [inline org.kde.ghostwriter.yml:24]:
```suggestion
```

this is not necessary

File: org.kde.ghostwriter.yml (hunk 5 above)
Comment (Justinzobel) [reply org.kde.ghostwriter.yml:77]: Yes this should be https://download.kde.org/stable/release-service/23.04.3/src/ghostwriter-23.04.3.tar.xz

**Comment by Aman9das** [comment]
bot, build org.kde.ghostwriter

**Comment by Aman9das** [comment]
bot, build org.kde.ghostwriter

File: org.kde.ghostwriter.yml [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+runtime: org.kde.Platform
+runtime-version: 5.15-22.08
+base: io.qt.qtwebengine.BaseApp
+base-version: 5.15-22.08
+sdk: org.kde.Sdk
+command: ghostwriter
+rename-icon: ghostwriter
+finish-args:
+- "--socket=wayland"
+- "--socket=x11"
```
Comment (travier) [inline org.kde.ghostwriter.yml:11]:
```suggestion
- "--socket=fallback-x11"
```

File: org.kde.ghostwriter.yml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+- "--socket=pulseaudio"
+- "--share=ipc"
+- "--share=network"
+- "--filesystem=home"
```
Comment (travier) [inline org.kde.ghostwriter.yml:15]: Any reason this is needed? Is the file portal not good enough for this app? If so, you should add a README with a small note explaining the reason here.

File: org.kde.ghostwriter.yml [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+- "--device=dri"
+- "--env=QTWEBENGINE_DISABLE_SANDBOX=1"
+modules:
+- name: cmark
+  buildsystem: cmake-ninja
+  builddir: true
+  config-opts:
+  - "-DCMAKE_BUILD_TYPE=Release"
```
Comment (travier) [inline org.kde.ghostwriter.yml:23]: I don't think this is needed

File: org.kde.ghostwriter.yml [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  sources:
+  - type: git
```
Comment (travier) [inline org.kde.ghostwriter.yml:25]: You should use tarballs as much as possible

File: org.kde.ghostwriter.yml [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    url: https://github.com/commonmark/cmark.git
+    tag: 0.30.3
+- name: cmark-gfm
+  buildsystem: cmake-ninja
+  builddir: true
+  config-opts:
+  - "-DCMAKE_BUILD_TYPE=Release"
+  sources:
+  - type: git
```
Comment (travier) [inline org.kde.ghostwriter.yml:34]: idem

File: org.kde.ghostwriter.yml [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    url: https://github.com/github/cmark.git
+    tag: 0.29.0.gfm.13
+- name: MultiMarkdown
+  buildsystem: cmake-ninja
+  builddir: true
+  config-opts:
+  - "-DCMAKE_BUILD_TYPE=Release"
```
Comment (travier) [inline org.kde.ghostwriter.yml:41]: idem

**Comment by Aman9das** [comment]
bot, build org.kde.ghostwriter

**Comment by hfiguiere** [comment]
it there a PR for the EOL rebase of the original app?

**Comment by travier** [comment]
The git repo sources should be replaced by archives with x-data-checker configs (https://github.com/flathub/flatpak-external-data-checker).

File: org.kde.ghostwriter.yml [hunk 13]
```diff
@@ -0,0 +1,58 @@
+app-id: org.kde.ghostwriter
+runtime: org.kde.Platform
+runtime-version: 5.15-22.08
+base: io.qt.qtwebengine.BaseApp
```
Comment (bbhtt) [inline org.kde.ghostwriter.yml:4]: Cleanup baseapp https://github.com/flathub/io.qt.qtwebengine.BaseApp/wiki

Comment (Justinzobel) [reply org.kde.ghostwriter.yml:4]: Ping for cleanup.

File: org.kde.ghostwriter.yml (hunk 10 above)
Comment (Aman9das) [reply org.kde.ghostwriter.yml:21]: Couldn't find any release tarball ...

**Comment by Aman9das** [comment]
I have been unable to find tarballs unfortunately.

**Comment by Aman9das** [comment]
bot, build org.kde.ghostwriter

**Comment by Justinzobel** [comment]
> I have been unable to find tarballs unfortunately.

https://download.kde.org/stable/release-service/24.02.1/src/ghostwriter-24.02.1.tar.xz

**Comment by Aman9das** [comment]
> > I have been unable to find tarballs unfortunately.
> 
> https://download.kde.org/stable/release-service/24.02.1/src/ghostwriter-24.02.1.tar.xz

Yeah only for pandoc and ghostwriter could find tarball. Building cmark, cmark-gfm and multimarkdown from source

**Comment by Justinzobel** [comment]
> > > I have been unable to find tarballs unfortunately.
> > 
> > 
> > https://download.kde.org/stable/release-service/24.02.1/src/ghostwriter-24.02.1.tar.xz
> 
> Yeah only for pandoc and ghostwriter could find tarball. Building cmark, cmark-gfm and multimarkdown from source

Ah my mistake.

**Comment by Aman9das** [comment]
Also I am a little bit confused between Qt6 and Qt5
According to https://github.com/flathub/io.qt.qtwebengine.BaseApp/wiki

Should I set the BaseApp at 6.7 to enable Qt6

File: org.kde.ghostwriter.yml [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+base-version: 5.15-22.08
```
Comment (bbhtt) [inline org.kde.ghostwriter.yml:5]: Why not use 5.15-23.08?

File: org.kde.ghostwriter.yml [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+  - "--device=dri"
+cleanup-commands:
+  - /app/cleanup-BaseApp.sh
+modules:
+  - name: cmark
+    buildsystem: cmake-ninja
+    builddir: true
+    sources:
+      - type: git
+        url: https://github.com/commonmark/cmark.git
```
Comment (bbhtt) [inline org.kde.ghostwriter.yml:24]: You can use auto generated tarballs https://github.com/commonmark/cmark/archive/refs/tags/0.31.0.tar.gz

File: org.kde.ghostwriter.yml (hunk 14 above)
Comment (Aman9das) [reply org.kde.ghostwriter.yml:5]: I'm trying out 6.7 rn

File: org.kde.ghostwriter.yml [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+        url: https://github.com/fletcher/MultiMarkdown-6.git
+        tag: 6.7.0
+        commit: 6b6155d5041a023d0de44adf182506c52202a03c
+        x-checker-data:
+          type: anitya
+          project-id: 131019
+          stable-only: true
+          tag-template: $version
+  - name: Pandoc
+    buildsystem: simple
```
Comment (bbhtt) [inline org.kde.ghostwriter.yml:59]: Why is this not being built from source?

File: org.kde.ghostwriter.yml (hunk 15 above)
Comment (Aman9das) [reply org.kde.ghostwriter.yml:24]: It's source code..

File: org.kde.ghostwriter.yml (hunk 14 above)
Comment (bbhtt) [reply org.kde.ghostwriter.yml:5]: Ok, the baseapp should be available soonish. I created the branch but Flathub has having some server issues that delays publishing.

File: org.kde.ghostwriter.yml (hunk 16 above)
Comment (Aman9das) [reply org.kde.ghostwriter.yml:59]: pandoc has release tarball with binary

File: org.kde.ghostwriter.yml (hunk 14 above)
Comment (Aman9das) [reply org.kde.ghostwriter.yml:5]: 6.6 if for qt6 but is giving error:

File: org.kde.ghostwriter.yml (hunk 15 above)
Comment (Aman9das) [reply org.kde.ghostwriter.yml:24]: Oh now I get it.

File: org.kde.ghostwriter.yml (hunk 16 above)
Comment (Aman9das) [reply org.kde.ghostwriter.yml:56]: also its complex to build

**Comment by Aman9das** [comment]
bot, build org.kde.ghostwriter

**Comment by Aman9das** [comment]
Ready now

**Comment by Aman9das** [comment]
I am familiar with only yaml. The original manifest was in yaml. I think it'd be fairly straightforward to convert to json by KDE contributors once it has been accepted into flathub. :)

**Comment by Justinzobel** [comment]
> It uses yml when in KDE we've mostly standarized on using json but IDK, now that i don;t use my custom script to increase versions and we have the x-chcker-data maybe it's not too bad to have a bit of mix on the format side

I much prefer YAML to JSON, however I'm OK working in both.

**Comment by travier** [comment]
For now we were using only JSON to keep things consistent. Ideally we would migrate them all in one go.
I'm not opposed to adding this one as is. I've not tested it however.

**Comment by Aman9das** [comment]
bot, build org.kde.ghostwriter

**Comment by Aman9das** [comment]
Converted to json

**Comment by travier** [comment]
Thanks. Is this a Qt5/KF5 app still or has it been ported to Qt6/KF6? We now have a qt6.6 & qt6.7 runtime.

**Comment by Aman9das** [comment]
bot, build org.kde.ghostwriter

reverted the qt6 build that fails

**Comment by Aman9das** [comment]
@travier @tsdgeos @Justinzobel how would app verification be possible. Ideally the app developers should maintain the Flatpak..

**Comment by Justinzobel** [comment]
App verification happens once it's on the store. The Flathub team handle assigning it to KDE and then we validate that it's our app on the Flathub site.

**Comment by travier** [comment]
I've clicked the verify button. It should be displayed as verified "soon".

**Pull request closed without merging**